File: voice_assistant/voice_listener/domain/service.py
import threading

from voice_listener.domain.model import Language, WakeWord
from voice_listener.domain.ports import AudioCapturer, AudioSpeaker, OrderHandler, Transcriber

_ORDER_TIMEOUT = 10          # seconds to wait for the user to start speaking
_ORDER_RETRIES = 2            # how many times to re-prompt before giving up
_ORDER_PAUSE_THRESHOLD = 2.0  # seconds of silence before the order phrase is considered done


class VoiceListenerService:
    def __init__(
        self,
        capturer: AudioCapturer,
        transcriber: Transcriber,
        order_handler: OrderHandler,
        speaker: AudioSpeaker,
        wake_word: WakeWord,
        language: Language,
    ) -> None:
        self._capturer = capturer
        self._transcriber = transcriber
        self._order_handler = order_handler
        self._speaker = speaker
        self._wake_word = wake_word
        self._language = language

    def wait_for_wake_word(self) -> str | None:
        """Block until the wake word is heard. Returns any inline order found in the same utterance."""
        print(f'Waiting for wake word "{self._wake_word.value}"...')
        while True:
            audio = self._capturer.capture(timeout=None, phrase_time_limit=8)
            if audio is None:
                continue
            text = self._transcriber.transcribe(audio, self._language)
            if text:
                print(f"Heard: {text!r}")
            if text and self._wake_word.matches(text):
                print("Wake word detected!")
                return self._wake_word.extract_order(text)

    def listen_for_order(self) -> str | None:
        """Listen for an order, retrying up to _ORDER_RETRIES times."""
        for attempt in range(_ORDER_RETRIES):
            self._speaker.beep()
            print("Listening for your order... speak now!")
            audio = self._capturer.capture(
                timeout=_ORDER_TIMEOUT,
                phrase_time_limit=None,
                pause_threshold=_ORDER_PAUSE_THRESHOLD,
            )
            if audio is None:
                if attempt < _ORDER_RETRIES - 1:
                    print("I didn't catch that. Please repeat your order.")
                continue
            text = self._transcriber.transcribe(audio, self._language)
            if text:
                return text
        print("No order detected.")
        return None

    def run(self) -> None:
        waiting_for_answer = False
        while True:
            if waiting_for_answer:
                order = self.listen_for_order()
            else:
                inline_order = self.wait_for_wake_word()
                order = inline_order or self.listen_for_order()
            if order:
                print(f"Order received: {order!r}")
                response, stop_melody, melody_thread = self._handle_with_melody(order)
                print(f"Claudito: {response}")
                interrupted = self._speak_interruptible(response, stop_melody, melody_thread)
                if interrupted:
                    waiting_for_answer = False
                else:
                    waiting_for_answer = bool(response and "?" in response.rstrip())
            else:
                waiting_for_answer = False

    def _handle_with_melody(self, order: str) -> tuple[str, threading.Event, threading.Thread]:
        """Call the order handler while playing a waiting melody. Returns the melody still running."""
        stop_event = threading.Event()
        melody_thread = threading.Thread(
            target=self._speaker.play_melody, args=(stop_event,), daemon=True
        )
        melody_thread.start()
        response = self._order_handler.handle(order)
        return response, stop_event, melody_thread

    def _speak_interruptible(self, response: str, stop_melody: threading.Event, melody_thread: threading.Thread) -> bool:
        """Speak response while listening for the wake word. Returns True if interrupted."""
        speak_thread = threading.Thread(
            target=self._speaker.speak,
            args=(response, self._language),
            kwargs={"on_playback_start": stop_melody.set},
            daemon=True,
        )
        speak_thread.start()
        melody_thread.join()

        self._capturer.set_echo_reference(self._speaker.get_echo_reference())
        interrupted = False
        try:
            while speak_thread.is_alive():
                audio = self._capturer.capture(timeout=1, phrase_time_limit=2)
                if audio is not None:
                    text = self._transcriber.transcribe(audio, self._language)
                    if text:
                        print(f"Heard during speech: {text!r}")
                    if text and self._wake_word.matches(text):
                        print("Wake word detected — interrupting speech.")
                        self._speaker.stop()
                        self.listen_for_order()
                        interrupted = True
                        break
        finally:
            self._capturer.set_echo_reference(None)

        speak_thread.join()
        return interrupted

File: voice_assistant/rust/src/domain/service.rs
use std::sync::{Arc, atomic::{AtomicBool, Ordering}};
use std::thread;

use crate::domain::model::{Language, WakeWord};
use crate::domain::ports::{AudioCapturer, AudioSpeaker, OrderHandler, Transcriber};

const ORDER_TIMEOUT_MS:       u64   = 10_000;
const ORDER_RETRIES:          usize = 2;
const ORDER_PAUSE_THRESHOLD:  u64   = 2_000;

pub struct VoiceListenerService {
    capturer:      Box<dyn AudioCapturer>,
    transcriber:   Arc<dyn Transcriber>,
    order_handler: Arc<dyn OrderHandler>,
    speaker:       Arc<dyn AudioSpeaker>,
    wake_word:     WakeWord,
    language:      Language,
}

impl VoiceListenerService {
    pub fn new(
        capturer:      Box<dyn AudioCapturer>,
        transcriber:   Arc<dyn Transcriber>,
        order_handler: Arc<dyn OrderHandler>,
        speaker:       Arc<dyn AudioSpeaker>,
        wake_word:     WakeWord,
        language:      Language,
    ) -> Self {
        Self { capturer, transcriber, order_handler, speaker, wake_word, language }
    }

    // ── public methods ────────────────────────────────────────────────────────

    pub fn wait_for_wake_word(&mut self) -> Option<String> {
        eprintln!("[listening for wake word \"{}\"]", self.wake_word.value);
        loop {
            let audio = self.capturer.capture(None, Some(8_000), None);
            let audio = match audio { Some(a) => a, None => continue };
            let text  = self.transcriber.transcribe(&audio, &self.language);
            if let Some(ref t) = text {
                eprintln!("[heard: {t:?}]");
                if self.wake_word.matches(t) {
                    return self.wake_word.extract_order(t);
                }
            }
        }
    }

    pub fn listen_for_order(&mut self) -> Option<String> {
        eprintln!("[listening for order]");
        for attempt in 0..ORDER_RETRIES {
            self.speaker.beep();
            let audio = self.capturer.capture(
                Some(ORDER_TIMEOUT_MS),
                None,
                Some(ORDER_PAUSE_THRESHOLD),
            );
            let audio = match audio {
                Some(a) => a,
                None => {
                    if attempt < ORDER_RETRIES - 1 {
                        eprintln!("I didn't catch that.");
                    }
                    continue;
                }
            };
            if let Some(t) = self.transcriber.transcribe(&audio, &self.language) {
                return Some(t);
            }
        }
        None
    }

    pub fn handle_with_melody(
        &mut self,
        order: &str,
    ) -> (String, Arc<AtomicBool>, thread::JoinHandle<()>) {
        let stop_signal   = Arc::new(AtomicBool::new(false));
        let stop_clone    = Arc::clone(&stop_signal);
        let speaker_clone = Arc::clone(&self.speaker);

        let melody_thread = thread::spawn(move || {
            speaker_clone.play_melody(stop_clone);
        });

        let response = self.order_handler.handle(order);
        (response, stop_signal, melody_thread)
    }

    pub fn run(&mut self) {
        println!("Voice Order Listener");
        println!("====================");
        println!("Press Ctrl+C to quit.\n");

        let mut waiting_for_answer = false;
        loop {
            let order = if waiting_for_answer {
                self.listen_for_order()
            } else {
                let inline = self.wait_for_wake_word();
                inline.or_else(|| self.listen_for_order())
            };

            if let Some(ref order_text) = order {
                println!("Order received: {order_text:?}");
                let (response, stop_melody, melody_thread) =
                    self.handle_with_melody(order_text);
                println!("Claudito: {response}");
                let interrupted =
                    self.speak_interruptible(&response, stop_melody, melody_thread);
                if interrupted {
                    waiting_for_answer = true;
                } else {
                    waiting_for_answer = response.trim_end().ends_with('?');
                }
            } else {
                waiting_for_answer = false;
            }
        }
    }

    pub fn speak_interruptible(
        &mut self,
        response:      &str,
        stop_melody:   Arc<AtomicBool>,
        melody_thread: thread::JoinHandle<()>,
    ) -> bool {
        let speaker_clone  = Arc::clone(&self.speaker);
        let response_owned = response.to_string();
        let lang_clone     = self.language.clone();
        let stop_clone     = Arc::clone(&stop_melody);

        let speak_handle = thread::spawn(move || {
            speaker_clone.speak(
                &response_owned,
                &lang_clone,
                Some(Box::new(move || stop_clone.store(true, Ordering::SeqCst))),
            );
        });

        // Wait for melody to stop (it stops when on_playback_start fires)
        let _ = melody_thread.join();

        self.capturer.set_echo_reference(self.speaker.get_echo_reference());

        let mut interrupted = false;
        while !speak_handle.is_finished() {
            let audio = self.capturer.capture(Some(1_000), Some(2_000), None);
            if let Some(ref a) = audio {
                if let Some(ref t) = self.transcriber.transcribe(a, &self.language) {
                    if self.wake_word.matches(t) {
                        self.speaker.stop();
                        interrupted = true;
                        break;
                    }
                }
            }
        }

        self.capturer.set_echo_reference(None);
        let _ = speak_handle.join();
        interrupted
    }
}
